### PR TITLE
Add more functions to window API

### PR DESCRIPTION
**Pull request by sebastiansam55**
Adds `get_window_geom`, `get_window_hex` and `get_window_list` to the window API. Allows for use of other commands like `import -window <hexid>` for getting screenshots of windows (for upcoming visgrep changes).

### CODE REVIEW

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: Tidy up the inside of this loop with a function maybe? Something that just returns the list of useful info?

You can use `@staticmethod` above the function declaration to avoid having to include API documentation, I think.

If you're really keen, getting some tests written where you can would be great. Even if you mock out the external bits to return the expected results, that would still be super useful.

**Review by BlueDrink9 — CHANGES_REQUESTED** [review]

File: lib/autokey/scripting/window.py
Comment (sebastiansam55) [reply lib/autokey/scripting/window.py]: I may have brought it up elsewhere but I think we should probably swap from using wmctrl to internal implementation because implementing reading output from the commandline is always going to be a pain to implement (and be easily broken if `wmctrl` changes anything about their output)

I say this understanding that the main thing we are lacking atm is developer time and re implementing existing features is not the best use of time

Comment (BlueDrink9) [reply lib/autokey/scripting/window.py]: yeah, reasonable point. One stop-gap to cover wmctrl changing output is to create some unit tests that assert its interface is what we expect it to be. Have you had a go with the testing system yet? It's very liberating once you get used to it.

Comment (BlueDrink9) [reply lib/autokey/scripting/window.py]: I've changed my mind about re-writing this loop (so ignore my comment about `@staticmethod`). Would still be nice to have some tests though.

I think the easiest way to start is to create `tests/scripting_api/test_window.py`. Copy the contents of `tests/scripting_api/test_macro.py` to get the relevant setup methods and imports, etc. (Feel free to change the copyright notice at the top to your name/username and email).

The [mock objects](https://docs.python.org/3/library/unittest.mock.html) and [patches](https://docs.python.org/3/library/unittest.mock.html#patch) that you can create are very useful. In this case, you might want a test body like 

```python
get_win_details = MagicMock(return_value=test_string)
with patch("self._run_wmctrl", get_win_details):
    output = engine.get_window_list()
    assert_equal(output, expected_output)
```

Where `test_string` is an example of what you might expect `self._run_wmctrl(["-l"])` to return in a real system, and `expected_output` is what you expect the function to return.

Create one of these for now, and when you've got the hang of it, I'll show you how to test multiple test strings in a single test. (Or play with the `@parameterize` decorator if you want to learn yourself).

Each test you write is a function in the file that starts with `test_`.

**Comment by sebastiansam55** [comment]
Rebased to develop and added center_window convenience method.

**Comment by BlueDrink9** [comment]
For a "convenience" method, it's a pretty long method! It's longer than most of the other top-level methods you've added so far. It would be good to come back and refactor it.

Also let me know how you are getting on with tests. Feel free to push them, either here or on a branch on your fork, if you want me to review them or help out

**Comment by sebastiansam55** [comment]
True, convenience more in the sense I wanted some way to center windows and it was surprisingly hard to find modern methods to accomplish this without writing code, even this method is still 100% reliant on wmctrl and xrandr (most online solutions recommend compiz or digging into dconf) But as long as I was going to write code for it it might as well be in AutoKey so others can take advantage of it.

RE testing;
I've taken a few swings at writing some and am making more process than I have previously! Your comments on what I'm doing are more than welcome, I'm sure that I'm not following all of the coding conventions around tests...

Also I annotated a few places in the `window.py` file that seem like they might make more sense as an enum like those used in `button.py` or `key.py`. More specifically the special methods for window selection (`":ACTIVE:", ":SELECT:"`) (line 30 in window.py) and the properties and actions wmctrl offers, as seen described in the comment for the `set_property` method (line 184ish)

**Comment by BlueDrink9** [comment]
I'm not sure how helpful I can be in terms of testing conventions; I'm still learning them myself. But I will try my best to at least pull you up to where I'm at!

**Comment by BlueDrink9** [comment]
The Enum question is an interesting one. It seems like a good thing in general, but I think this is a special case because these are public API methods. We want them to be callable without anything extra being imported, so it seems like using strings instead of enums is reasonable here. If they were being used internally (which they shouldn't be?) then I would agree. Internal use = enum, but external/docs should use the equivalent string.

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: This is way less clear than the previous version. Don't hide details under complicated python syntax just to save a few lines, IMO. Especially in a project like this, where maintainers may not know python well.

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: Can you comment this regex?

Also, for regexes especially, tests are great for two reasons:
1. Easy enough to write, just have a bunch of examples and ensure it matches them and not others. Check `test_macro` for examples of regex tests.
2. Acts as a way to self-document exactly what the regex does.

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: This is inappropriate in a docstring comment. Stick it outside the docstring.

As to why it is included: No idea

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: Write a comment that breaks this regex down into its components please. Also things like examples of input it expects, and example of match/output. (Doing this in tests would also be a good idea; point people to the test in a comment here).

Also consider making your regexes specific constants, [like is done elsewhere in the codebase](https://github.com/autokey/autokey/blob/4b87b32452cd41d2b98341db220e99cf4f9e34d6/lib/autokey/model.py#L35)

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: What do you mean by gravity, front end, back end? Can you give an example in this comment? (also put it the line above, not inline)

File: tests/scripting_api/test_window.py
Comment (BlueDrink9) [inline tests/scripting_api/test_window.py]: [Don't need multiple `with`s](https://github.com/autokey/autokey/blob/29a8b86ecc0ed3e210b1845e2599b94a305aaf03/tests/engine_helpers.py#L43)

**Review by BlueDrink9 — CHANGES_REQUESTED** [review]

File: tests/scripting_api/test_window.py
Comment (BlueDrink9) [inline tests/scripting_api/test_window.py:177]: There is a way, but it isn't much difference because you still have to pass it in as an argument. ([Example here](https://github.com/autokey/autokey/blob/29a8b86ecc0ed3e210b1845e2599b94a305aaf03/tests/scripting_api/test_engine.py#L82) [fixture defined here](https://github.com/autokey/autokey/blob/29a8b86ecc0ed3e210b1845e2599b94a305aaf03/tests/engine_helpers.py#L35)) This is fine as a one-liner.

That being said, I'm not an expert so I might have missed a better way to do it. Pytest's docs on fixtures are a little confusing.

**Review by BlueDrink9 — COMMENTED** [review]
The tests you have there are looking really good! I'm impressed and heartened, they're definitely better than my early attempts.

**Comment by BlueDrink9** [comment]
Rebased atop develop on your behalf and force-pushed after merging of #494

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: This branch (`filter_desktop is not None`) doesn't currently have test coverage. Would be nice to get something in here.

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [inline lib/autokey/scripting/window.py]: How does this function compare to `get_active_geometry` (line 201)? Can you please deduplicate? (Would be nice to have your tests converted for `get_active_geometry` as well)

File: lib/autokey/scripting/window.py
Comment (sebastiansam55) [reply lib/autokey/scripting/window.py]: Gravity is a weird thing that is in the X11 window manager spec but appears to not be used/respected by at least the WM I use (whatever the Ubuntu 20.04 defaults to).

You can read more about it here https://specifications.freedesktop.org/wm-spec/wm-spec-1.3.html#Window%20Geometry

From reading about it, gravity seems like it would be an ideal way to center windows using `wmctrl`, my understanding is that it changes the "anchor" point where the window is placed, defaulting to the top left hand corner as one might expect, and each different value choosing a different point on either the midpoints or corners of the sides or the center itself

I may be reading it wrong, there are lots of words there lol. There seems to be some difference between the "client window" and the "frame window" but I'm not sure what it is referring to. 

I would include gravity in the return values for this function, as it does seem relevant to window geometry but the pre-existing `get_active_geometry` only returns the offset x,y and width/height. So for consistency I choose to discard it.

Not aware of anyone ever asking/wondering where it is, and from my googling of it it does not come up a lot.

Comment (sebastiansam55) [reply lib/autokey/scripting/window.py]: It should be considered that the two methods (`get_window_geom` and `get_active_geometry`) are more or less equivalent, I'd advocate that we should make `get_active_geometry` as deprecated and change up `get_window_geom` to accomplish the same thing (would get the active window by default, and take title/hex id as optional args)

The two functions are doing essentially the same thing (calling wmctrl and parsing the output), just `get_active` is limited to only getting the active window's geometry obviously.

File: tests/scripting_api/test_window.py
Comment (sebastiansam55) [reply tests/scripting_api/test_window.py]: Yeah I figured that there had to be some way to do it but was not aware of the syntax, neat!

**Comment by josephj11** [comment]
Just wondering: Is there a character class for hex? If there is, would it be better to use in the regexes in case someone's strange locale's collating sequence contains some surprises?

**Comment by sebastiansam55** [comment]
> Just wondering: Is there a character class for hex? If there is, would it be better to use in the regexes in case someone's strange locale's collating sequence contains some surprises?

I'm not sure I did try `[[:xdigit:]]` but python didn't like it and according to the website I was using (regex101.com):
```
[[:xdigit:]]

Matches hexadecimal digits, case insensitive. Equivalent to [0-9a-fA-F].
```

File: lib/autokey/scripting/window.py
Comment (BlueDrink9) [reply lib/autokey/scripting/window.py]: Rather than deprecate, how about you just convert it into a wrapper than calls your func?

**Comment by BlueDrink9** [comment]
> but python didn't like it and according to the website I was using (regex101.com)

[Python uses similar regex to perl flavour](https://docs.python.org/3/library/re.html#module-re), which is likely to be relevant when it comes to choosing flavour of regex on the website.

**Comment by sebastiansam55** [comment]
> > but python didn't like it and according to the website I was using (regex101.com)
> 
> I suggest

I think you may have sent this message before completing it haha

**Comment by sebastiansam55** [comment]
@BlueDrink9 
* I see in the original `get_active_geometry` there are the following lines; (https://github.com/autokey/autokey/blob/develop/lib/autokey/scripting/window.py#L216)

```
# return [int(x) for x in output]
return list(map(int, output))
```

They seem to be equivalent, and I use the commented out method in other methods contained in this commit, should I convert them to use the other "newer" (presumably) method?

* Also in `_run_wmctrl` the newline is stripped via character indexing, instead of something like `.strip()` should this be updated to move away from.

* I would also like to say that I might add a `by_hex` flag to many (if not all of the existing functions) in order to address the issues raised in issue #306 . Because they are right, the way it works right now if there is umore than one window with the same name you'd only ever be able to select/act upon the "first" one in the list. 

The flag would be set by default to false in order to maintain backward compatibility of course.

**Comment by BlueDrink9** [comment]
Yes, your points are all good ones. I think `return list(map(int, output))` is the better way, because it's easier to look up what `map` does than Google the name for list comprehension's syntax.

File: tests/scripting_api/test_window.py
Comment (BlueDrink9) [reply tests/scripting_api/test_window.py:177]: Might be able to use the example I linked, but make it an [autouse fixture](https://docs.pytest.org/en/6.2.x/fixture.html#autouse-fixtures-fixtures-you-don-t-have-to-request) (`@pytest.fixture(autouse=True)`)

**Comment by BlueDrink9** [comment]
Great round of changes, nice work

**Review by BlueDrink9 — APPROVED** [review]

**Pull request merged**
